lib/swig/GeNNModel.py

```diff
@@ -0,0 +1,934 @@
+            else:
+                var = self.synapsePopulations[popName].vars[varName]
+        else:
+            var = self.neuronPopulations[popName].vars[varName]
+
+        var.view[mask] = vals
+        self.pushStateToDevice( popName )
+
+
+    def initializeSpikesOnDevice( self, popName, mask, targets, counts ):
+        """Set spike counts and targets for the given population and push them to the device
+        
+        Note: shapes of mask, targets and counts must be the same
+
+        Args:
+        popName -- name of the population for which the spikes must be set
+        mask    -- list with source neurons ids
+        targets -- list with target neuron ids
+        counts  -- list with number of spikes for source neurons
+        """
+        if popName not in self.neuronPopulations:
+            raise ValueError( 'Failed to initialize variable "{0}": \
+                    population "{1}" does not exist'.format( varName, popName ) )
+        self.neuronPopulations[popName].spikes[mask] = targets
+        self.neuronPopulations[popName].spikeCount[mask] = counts
+        self.pushSpikesToDevice( popName )
+
+
+    def build( self, pathToModel = "./" ):
+
+        """Finalize and build a GeNN model
+        
+        Keyword args:
+        pathToModel -- path where to place the generated model code. Defaults to the local directory.
+        """
+
+        if self._built:
+            raise Exception("GeNN model already built")
+        self._pathToModel = pathToModel
+
+        for popName, popData in self.synapsePopulations.items():
+            if popData.sparse:
+                popData.pop.setMaxConnections( popData.size )
+
+        self._model.finalize()
+        #  if not self._cpuOnly:
+        #      pg.chooseDevice(self._model, pathToModel, self._localhost)
+        pg.finalize_model_runner_generation(self._model, self._pathToModel, self._localhost)
+
+        check_call( ['make', '-C', path.join( pathToModel, self.modelName + '_CODE' ) ] )
+        
+        self._built = True
+
+    def load( self ):
+        """import the model as shared library and initialize it"""
+        if not self._built:
+            raise Exception( "GeNN model has to be built before running" )
+        
+        self._slm.open( self._pathToModel, self.modelName )
+
+
+        self._slm.allocateMem()
+        self._slm.initialize()
+        
+        if self._scalar == 'float':
+            self._T = self._slm.assignExternalPointerSingle_f('t')
+        if self._scalar == 'double':
+            self.T = self._slm.assignExternalPointerSingle_d('t')
+        if self._scalar == 'long double':
+            self._T = self._slm.assignExternalPointerSingle_ld('t')
+        self._TS = self._slm.assignExternalPointerSingle_ull('iT')
+
+        for popName, popData in self.neuronPopulations.items():
+            self._slm.initNeuronPopIO( popName )
+            self.pullStateFromDevice( popName )
```

Comment:
How come you're **pulling** here? Also for synapse groups etc

Reply:
Ah this is from one of the previous versions. There was an option to provide a mask and rewrite just few of selected variables. In combination with `VarMode::LOC_HOST_DEVICE_INIT_DEVICE` pulling was necessary, but it was complicating the module too much. Now it is unnecessary.

Reply:
After double checking these lines, I believe pulling is still needed, since pull/push affects **all** variables of the population. So, if values need to be pushed only to one variable, without pulling other variables will get whatever garbage was in the memory at the time of allocation.

A check whether **any** variables need to be pushed could avoid unnecessary pulling/pushing.

Reply:
Ahh ok! So firstly you should be doing manual variable initialisation **between** the calls to the auto-generated ``initialize()`` function and the auto-generated ``initializeMODEL_NAME()`` function. That way, assuming that:
1. ``autoInitSparseVars=True`` (as it should be)
2. The variables you are manually initialising are marked as uninitialised (which it looks liek they are)
then everything that needs to be copied will be copied by ``initializeMODEL_NAME()`` function - you shouldn't need to do any manual pushing or pulling

Reply:
I have checked the `initMODEL_NAME()` function and it only has `initializeSparseArrays()` in it. I tried providing a list of initial voltages and they do not appear in `initializeDevice()` as it should be, so `unintializedVar()` must be working, at least partially...

Reply:
How strange, if you have set ``GENN_PREFERENCES::autoInitSparseVars`` to true there **should** be a call to ``copyStateToDevice(true);`` before ``initializeSparseArrays()`` is called.

Reply:
OK, it seems I have overlooked that I was setting `autoInitSparseVars=false` elsewhere. Now the `initMODEL_NAME()` has `copyStateToDevice(true)`, `initializeAllSparseArrays()` and `initiazileSparseDevice()` in it. Sorry for misleading you.

However, `copyStateToDevice(true)` has no effect, because when it comes to individual `cudaMemcpy`s they are all wrapped in `if(!hostInitializedOnly) {` and therefore are never executed.

Also it seems that if the call were `copyStateToDevice(false)`, it would overwrite the variables initialized in `initializeDevice()` which in turn is called by `initialize()`.

It feels like I am missing some crucial bit somewhere...

Reply:
yeap, `VarMode::LOC_HOST_DEVICE_INIT_HOST` was the missing bit, thanks.